src/approaches/pg3_approach.py

```diff
@@ -0,0 +1,429 @@
+
+        def get_successors(ldl: _S) -> Iterator[Tuple[_A, _S, float]]:
+            for op in search_operators:
+                for i, child in enumerate(op.get_successors(ldl)):
+                    yield (op, i), child, 1.0  # cost always 1
+
+        if CFG.pg3_search_method == "gbfs":
+            path, _ = utils.run_gbfs(
+                initial_state=self._current_ldl,
+                check_goal=lambda _:
+                False,  # Terminate only after max expansions
+                get_successors=get_successors,
+                heuristic=heuristic,
+                max_expansions=CFG.pg3_gbfs_max_expansions,
+                lazy_expansion=True)
+
+        elif CFG.pg3_search_method == "hill_climbing":
+            path, _, _ = utils.run_hill_climbing(
+                initial_state=self._current_ldl,
+                check_goal=lambda _: False,  # Terminate when no improvement
+                get_successors=get_successors,
+                heuristic=heuristic,
+                enforced_depth=CFG.pg3_hc_enforced_depth)
+
+        else:
+            raise NotImplementedError("Unrecognized pg3_search_method "
+                                      f"{CFG.pg3_search_method}.")
+
+        # Save the best seen policy.
+        self._current_ldl = path[-1]
+        logging.info(f"Keeping best policy:\n{self._current_ldl}")
+        save_path = utils.get_approach_save_path_str()
+        with open(f"{save_path}_None.ldl", "wb") as f:
+            pkl.dump(self._current_ldl, f)
+
+    def learn_from_offline_dataset(self, dataset: Dataset) -> None:
+        # First, learn NSRTs.
+        self._learn_nsrts(dataset.trajectories, online_learning_cycle=None)
+        # Now, learn the LDL policy.
+        self._learn_ldl(dataset.trajectories)
+
+    def load(self, online_learning_cycle: Optional[int]) -> None:
+        load_path = utils.get_approach_load_path_str()
+        with open(f"{load_path}_{online_learning_cycle}.ldl", "rb") as f:
+            self._current_ldl = pkl.load(f)
+
+    def _create_search_operators(self) -> List[_PG3SearchOperator]:
+        search_operator_classes = [
+            _AddRulePG3SearchOperator,
+            _AddConditionPG3SearchOperator,
+        ]
+        preds = self._get_current_predicates()
+        nsrts = self._get_current_nsrts()
+        return [cls(preds, nsrts) for cls in search_operator_classes]
+
+    def _create_heuristic(
+            self, trajectories: Sequence[LowLevelTrajectory]) -> _PG3Heuristic:
+        preds = self._get_current_predicates()
+        nsrts = self._get_current_nsrts()
+        ground_atom_trajs = utils.create_ground_atom_dataset(
+            trajectories, preds)
+        heuristic_name_to_cls: Dict[str, TypingType[_PG3Heuristic]] = {
+            "policy_guided": _PolicyGuidedPG3Heuristic,
+            "policy_evaluation": _PolicyEvaluationPG3Heuristic,
+            "plan_comparison": _DemoPlanComparisonPG3Heuristic,
```

Comment:
rename this to `demo_plan_comparison` since PlanComparisonPG3Heuristic is its own thing (it's an abstract class)